### PR TITLE
Data sub patch

### PR DESCRIPTION
**Pull request by austinpaine**
Fixed data leaking before user logs in.  Created subscription manager to handle subscribing users to each data source as they add them to their dashboard instead of sending all the data all of the time.

### CODE REVIEW

File: src/clj/dashboard_clj/components/websocket.clj [hunk 1]
```diff
@@ -67,3 +81,129 @@
                          :handler           client-ev-handler
                          :webserver-adapter webserver-adapter
                          :options           options}))
+
```
Comment (austinpaine) [inline src/clj/dashboard_clj/components/websocket.clj:84]: Ignore huge comment below.  left it in for future reference if we need to see how some of these messages/events are structured

File: src/cljs/dashboard_clj/core.cljs [hunk 2]
```diff
@@ -14,9 +14,11 @@
...
 
 (defn connect-to-data-sources []
-  (let [{:keys  [chsk ch-recv send-fn state]} (sente/make-channel-socket! "/chsk" {:type :auto})]
+  (let [{:keys  [chsk ch-recv send-fn state]} (sente/make-channel-socket! "/chsk"
+                                                                          {:client-id @(rf/subscribe [:get-current-user])}
+                                                                          {:type :auto})]
 
-    (asyncm/go-loop []
+    (async/go-loop []
       (let [{:keys [event id ?data send-fn]} (async/<! ch-recv)]
```
Comment (austinpaine) [inline src/cljs/dashboard_clj/core.cljs:23]: noticed this typo that was in master (asyncm vs async), surprised the compiler didnt catch this, and it seemed to function as normal?  Changed to async, but no noticeable effects either way.  Strange

File: src/cljs/vanilla/core.cljs [hunk 3]
```diff
@@ -79,21 +78,6 @@
```
Comment (austinpaine) [inline src/cljs/vanilla/core.cljs:81]: This has been pulled out of core and moved to update-layout

File: src/cljs/vanilla/update_layout.cljs [hunk 4]
```diff
@@ -64,8 +86,9 @@
                                               :options ((:options conversion) options)))
                                           data)]
 
-                 ;(prn ":set-layout CONVERTED: " converted-data
-                 ;     "///// nextid: " highestNextid)
+                 ;(prn ":set-layout CONVERTED:  " converted-data)
+
+                 (data-source-subscribe (mapv #(:data-source %) converted-data))
```
Comment (austinpaine) [inline src/cljs/vanilla/update_layout.cljs:85]: the new data-source-subscribing happens every time the layout is updated

File: src/cljs/vanilla/widgets/pie_chart.cljs [hunk 5]
```diff
@@ -46,10 +46,10 @@
...
-    (prn "pie/convert-name-y " chart-type
-      " //// (data) " data
-      " //// (:data) " (get-in data [:data :series 0 :data])
-      " //// (ret) " ret)
+    ;(prn "pie/convert-name-y " chart-type
+    ;  " //// (data) " data
+    ;  " //// (:data) " (get-in data [:data :series 0 :data])
+    ;  " //// (ret) " ret)
 
     ret))
```
Comment (austinpaine) [inline src/cljs/vanilla/widgets/pie_chart.cljs:55]: noticed this print that got missed at some point and commented it out for cleanliness

**Review by cawasser — CHANGES_REQUESTED** [review]
Need to add some processing to basic-widget/[:remove-widget]-handler which should remove the "unused" data-source(s) from re-frame just to keep things tidy. I noticed that deleting a widget does NOT remove the data-source. Assuming the server would keep sending updates too, but didn't check for sure.

Also, since it is possible to subscribe to the same source from multiple widgets, we probably need to keep track of how many times a given source is subscribed and only remove it when the LAST widget is removed (count drops to 0) AND tell the server the client is no longer interested so it stops sending.

**Comment by austinpaine** [comment]
Oooh good catches, will implement

File: src/cljs/vanilla/widgets/basic_widget.cljs [hunk 6]
```diff
@@ -23,7 +28,8 @@
   :remove-widget
   (fn-traced [db [_ widget-id]]
     (delete-widget widget-id)
-    (assoc db :widgets (remove #(= (:key %) widget-id) (:widgets db)))))
+    (assoc db :widgets (remove #(= (:key %) widget-id) (:widgets db))
+              :data-sources (dissoc (:data-sources db) (get-data-source @(rf/subscribe [:active-widgets-by-user]) widget-id)))))
```
Comment (cawasser) [inline src/cljs/vanilla/widgets/basic_widget.cljs:32]: Interesting. Even without reference counting, a "removed" subscription will be resubscribed if another widget is still subscribing. It looks like this happens a part of layout management. We'll want to keep an eye on if this causes any thrashing.

**Review by cawasser — APPROVED** [review]

File: src/cljs/vanilla/widgets/basic_widget.cljs (hunk 6 above)
Comment (austinpaine) [reply src/cljs/vanilla/widgets/basic_widget.cljs:32]: You're quick!  Yep was just going to explain that it still worked based on how widgets get sent over if there are multiple widgets with the same data source.  Besides a flash as the new data updates get sent, no noticeable issues.  Might be worth a check once we turn scheduling back on eventually.

**Review by gpierce3 — CHANGES_REQUESTED** [review]
Noticed two things:

1. Looking at app-db when running in dev mode.  Picked two widgets.  Logged out.  data sources still showed in :datasources.  Is that what we want?

2. I picked a spectrum trace widget and then a carousel.  For kicks, spun thru the carousel.  The second widget was blank except for the chart title 'Sat2 Power'.  I then deleted the spectrum trace.  Now none of the chart info shows in carousel, just chart name.  It still 'knows' there are three widgets in the carousel. This happened in dev/jar/docker.  Went back to master branch and retried in dev mode.  Worked okay.

**Review by cawasser — COMMENTED** [review]
I added a fix for the (hard-coded) subscriptions for Carousel and WorldWind. Otherwise we get a feature regression (things stop working). This should also address gpierce3 comment #2 above.

Please double-check my work here.

**Comment by austinpaine** [comment]
Oh good catch with the logout data-sources George.  I had cleanup setup for when the user closed the tab (websocket connection close), but not logout.  Went down too hard of a path for a bit fixing this, but then realized i can just dissoc the data-sources and services easily in the logout event.  Done!

As for carousel, thats because the "data-source" for carousel is entirely fake and mocked up right now.  I hadnt tested this functionality with that widget because i didnt expect it to work at all.  I just took on the automatic carousel widget task today which will cover getting these subscriptions sorted out for carousel as well.

**Comment by austinpaine** [comment]
Thanks Chris for sticking in that quick fix!  Just looked it over and tested it and seems to work out great.  Will deal with getting it sorted out proper with the current carousel task

**Comment by cawasser** [comment]
We'll need to make some DB updates to add support for the "hard-coded" subscriptions for WW and Carousel, they aren't currently exposed that way. This will set things up for the "add a source" feature coming soon.

**Review by chadarm — APPROVED** [review]
Seems like a lot of good moves made in this PR, makes everything feel more organized. 
Great work with commenting and explaining functions, super easy to follow what's happening.

File: src/clj/dashboard_clj/components/websocket.clj [hunk 7]
```diff
@@ -67,3 +81,129 @@
...
+
+(comment
```
Comment (cawasser) [inline src/clj/dashboard_clj/components/websocket.clj:86]: should probably add a "real" comment describing what/why this is here

Comment (austinpaine) [reply src/clj/dashboard_clj/components/websocket.clj:86]: added

**Review by gpierce3 — APPROVED** [review]

**Comment by austinpaine** [comment]
Merged with latest master branch after live demo merge.  Fixed the only two conflicts.  Ready to merge.

File: src/cljs/vanilla/mapping/layer_management.cljs [hunk 8]
```diff
@@ -104,7 +104,8 @@
   ; TODO: this is a hack for the following hack (does NOT unsubscribe to sources when widget closes)
   (ds/data-source-subscribe [:beam-location-service :terminal-location-service])
```
Comment (cawasser) [inline src/cljs/vanilla/mapping/layer_management.cljs]: need to update the subscriptions to include :ka-beam-location-service

File: src/cljs/vanilla/widgets/carousel.cljs [hunk 9]
```diff
@@ -295,10 +295,10 @@
   ; TODO: this is a hack for the following hack (does NOT unsubscribe to sources when widget closes)
   (ds/data-source-subscribe [:spectrum-traces :power-data])
```
Comment (cawasser) [inline src/cljs/vanilla/widgets/carousel.cljs]: need to update the subscriptions to be [:channel-power-1000-service :channel-power-2000-service :channel-power-3000-service :channel-power-4000-service]

**Pull request merged**